src/Draco.Compiler.Api/Syntax/ParseTree.cs

```diff
@@ -42,10 +46,234 @@ public abstract partial class ParseTree
+        // The first token just gets it's content printed
+        // That ignores the leading trivia, trailing will only be printed if there are following tokens
+        Debug.Assert(tokensEnumerator.MoveNext());
+        var firstToken = tokensEnumerator.Current;
+        sb.Append(firstToken.Text);
+        lastTrailingTrivia = firstToken.TrailingTrivia;
+        while (tokensEnumerator.MoveNext())
+        {
+            var token = tokensEnumerator.Current;
+            // Last trailing trivia
+            foreach (var t in lastTrailingTrivia) sb.Append(t.Text);
+            // Leading trivia
+            foreach (var t in token.LeadingTrivia) sb.Append(t.Text);
+            // Content
+            sb.Append(token.Text);
+            // Trailing trivia
+            lastTrailingTrivia = token.TrailingTrivia;
+        }
+        return sb.ToString();
+    }
+
+    protected virtual Position ComputePosition()
+    {
+        // For simplicity we try to propagte to the first token
+        var firstToken = this.Tokens.FirstOrDefault();
+        if (firstToken is not null) return firstToken.Position;
+        // If we can't, we try to do so with the end of the last token
+        var precedingToken = this.GetPrecedingToken(this);
+        if (precedingToken is not null) return precedingToken.Range.End;
+        // Otherwise just default to 0, 0
+        return new(0, 0);
+    }
+
+    private Range ComputeRange()
+    {
+        var start = this.Position;
+        var end = StepPositionByText(start, this.ToString());
+        return new(start, end);
+    }
+
+    protected virtual IEnumerable<Token> GetTokens()
+    {
+        foreach (var child in this.Children)
+        {
+            foreach (var t in child.Tokens) yield return t;
+        }
+    }
+
+    private Token? GetPrecedingToken(ParseTree tree)
+    {
+        var preceding = null as Token;
+        foreach (var child in this.Children)
+        {
+            if (ReferenceEquals(child.Green, tree.Green)) break;
+            preceding = child.Tokens.LastOrDefault() ?? preceding;
+        }
+        if (preceding is not null) return preceding;
+        if (this.Parent is not null) return this.Parent.GetPrecedingToken(tree);
+        return null;
+    }
+
+    protected Token? GetPrecedingToken(Token token)
+    {
+        var preceding = null as Token;
+        foreach (var t in this.Tokens)
+        {
+            if (ReferenceEquals(t.Green, token.Green)) break;
+            preceding = t;
+        }
+        if (preceding is not null) return preceding;
+        if (this.Parent is not null) return this.Parent.GetPrecedingToken(token);
+        return null;
+    }
+
+    // NOTE: This might be a good general utility somewhere else?
+    private static Position StepPositionByText(Position start, string text)
+    {
+        var currLine = start.Line;
+        var currCol = start.Column;
+        for (var i = 0; i < text.Length; ++i)
```

Comment:
Why not Span\<char\>.EnumerateLines here or a SequenceReader\<char\> to do this ?

Reply:
It makes the position tracking logic a bit trickier to read. I understand that it eliminates the need to parse the newlines, but it takes me less to do that than to think about where to reset the column to 0 with a line-by-line iterator. I think the line-by-line iterator would even need an edge-case to handle the first line.

Reply:
While I understand for EnumerateLines, SequenceReader<char> track the position, it provide an AdvanceTo which would allow to quickly advance to a newline char, then do the usual logic.